wandb/sdk/mailbox/wait_with_progress.py

```diff
@@ -0,0 +1,135 @@
+from __future__ import annotations
+
+import time
+from typing import Any, Callable, Coroutine
+
+from wandb.proto import wandb_internal_pb2 as pb
+from wandb.sdk.lib import asyncio_compat
+
+from . import handles
+
+
+def wait_with_progress(
+    handle: handles.MailboxHandle,
+    *,
+    timeout: float | None,
+    progress_after: float,
+    display_progress: Callable[[], Coroutine[Any, Any, None]],
+) -> pb.Result:
+    """Wait for a handle, possibly displaying progress to the user.
+
+    Args:
+        handle: The handle to wait for.
+        timeout: A number of seconds after which to raise a TimeoutError,
+            or None if this should never timeout.
+        progress_after: A number of seconds after which to start the
+            progress_loop callback. Starting the callback creates a thread
+            and starts an asyncio loop, so we want to avoid doing it if
+            the handle is resolved quickly.
+        display_progress: An asyncio function that displays progress to
+            the user. This function is executed on a new thread and cancelled
+            if the timeout is exceeded.
+
+    Returns:
+        The result, if it is received before the timeout.
+
+    Raises:
+        TimeoutError: If the overall timeout expires.
+        HandleAbandonedError: If the handle becomes abandoned.
+        Exception: Any exception from the display function is propagated.
+    """
+    return wait_all_with_progress(
+        [handle],
+        timeout=timeout,
+        progress_after=progress_after,
+        display_progress=display_progress,
+    )[0]
+
+
+def wait_all_with_progress(
+    handle_list: list[handles.MailboxHandle],
+    *,
+    timeout: float | None,
+    progress_after: float,
+    display_progress: Callable[[], Coroutine[Any, Any, None]],
+) -> list[pb.Result]:
+    """Wait for multiple handles, possibly displaying progress to the user.
+
+    Arguments and exceptions are the same as for `wait_with_progress`.
+
+    Returns:
+        A list where the Nth item is the result of the Nth handle.
+    """
+    if not handle_list:
+        return []
+
+    if timeout is not None and timeout <= progress_after:
```

Comment:
could you add more comment about this code path vs the progress with timeout

Reply:
I considered writing a comment like

> When timeout <= progress_after, we will never show progress.

but this seems obvious. Do you think it would be useful, or is there something else I should elaborate on?

Reply:
no, now that i read it again, i agree this is fine